This notebook begins the modeling phase of our Facebook Metrics project.
All preprocessing steps have already been completed in preprocessing.ipynb, including:

Handling missing values , removing redundant / highly correlated columns, encoding categorical variables, scaling numerical features, saving the cleaned dataset to processed/facebook_cleaned.csv

In this notebook, we will:

- Load the cleaned dataset
- Define our target variable (y) and features (X)
- Create a train/test split
- Train a baseline model
- Evaluate performance
- Review feature importance

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [ ]:
#Loading the cleaned dataset

data = pd.read_csv("../processed/facebook_cleaned.csv")

In [ ]:
#Train/ Test Split 
#We chose 'like' because 'like' is the most stable/ computable metric to evaluate. 


X = data.drop(columns=['like'])
y = data['like']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

#Our supervised model is trying to predict the number of likes a Facebook post receives based on all the other features (reach & impression features, comments, shares, post type, and post metadata features).

**Random Forest Regressor Model**

In [ ]:
#We use RandomForestRegressor model first as a baseline because our data has mixed feature types and our variable 'like' is a nonlinear feature

from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

In [8]:
#Then Evaluate

from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print("MAE:", mae)
print("RMSE:", rmse)
print("R^2:", r2)


MAE: 0.19444093780244479
RMSE: 0.4610121390176264
R^2: 0.5164342223779319


**Results**

Our results show us how well does the model predict the number of likes a post will get. 
MAE: 0.1944 , on average our model's predictions are off by 0.19 units of our target variable. Our model then is reasonably accurate but not perfect.

RMSE: 0.4610 , model occassionally can ,ale larger errors than the MAE suggests. Some posts may be harder to predit or there may be outliers or unusal posts (viral posts, paid videos, paid posts, etc.)

R^2: 0.5164 , shows us how much variation our target variable model can explain. Which is about half of the variation in likes. 